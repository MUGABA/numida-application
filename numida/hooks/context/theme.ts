export type Theme = {
  background: string;
  primary: string;
  secondary: string;
  text: string;
  buttonText: string;
  border: string;
};

export const lightTheme: Theme = {
  background: "#ffffff",
  primary: "#30C2E3",
  secondary: "#D4FAF5",
  text: "",
  buttonText: "",
  border: "",
};

export const darkTheme: Theme = {
  background: "#121212",
  primary: "#bb86fc",
  secondary: "#03dac6",
  text: "#ffffff",
  buttonText: "#000000",
  border: "#333333",
};
