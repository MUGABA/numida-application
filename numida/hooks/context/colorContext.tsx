// ThemeContext.tsx
import React, { createContext, ReactNode, useContext, useState } from "react";
import { darkTheme, lightTheme, Theme } from "./theme";

interface ThemeContextProps {
  theme: Theme;
  toggleTheme: () => void;
}

const ThemeContext = createContext<ThemeContextProps | undefined>(undefined);

export const ColorThemeProvider: React.FC<{ children: ReactNode }> = ({
  children,
}) => {
  const [theme, setTheme] = useState<Theme>(lightTheme);

  const toggleTheme = () => {
    setTheme((prevTheme) =>
      prevTheme === lightTheme ? darkTheme : lightTheme
    );
  };

  return (
    <ThemeContext.Provider value={{ theme, toggleTheme }}>
      {children}
    </ThemeContext.Provider>
  );
};

export const useColorTheme = (): ThemeContextProps => {
  const context = useContext(ThemeContext);
  if (!context) {
    throw new Error("useTheme must be used within a ThemeProvider");
  }
  return context;
};
