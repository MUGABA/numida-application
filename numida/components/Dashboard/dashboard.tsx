import ButtonContainer from "@/components/custom/ButtonContainer";
import Card from "@/components/custom/Card/Card";
import { useColorTheme } from "@/hooks/context/colorContext";
import { gql, useQuery } from "@apollo/client";
import { useRouter } from "expo-router";
import React from "react";
import { View } from "react-native";
import ResponseGeneric from "../Response/ResponseGeneric";

type Props = {};

type LoanProduct = {
  id: number;
  name: string;
  interestRate: number;
  maximumAmount: number;
};

type LoanProducts = {
  loanProducts: LoanProduct[];
};

const GET_LOAN_PRODUCTS = gql`
  query {
    loanProducts {
      id
      name
      interestRate
      maximumAmount
    }
  }
`;

const Dashboard = (props: Props) => {
  const router = useRouter();
  const { theme } = useColorTheme();

  const { loading, error, data } = useQuery<LoanProducts>(GET_LOAN_PRODUCTS);

  const handleRedirectTOApplyForLoanScreen = () => {
    router.push("/loanApplication");
  };

  const loanProducts = !loading && data?.loanProducts;

  if (error) {
    return (
      <ButtonContainer
        title="Loan Application Dashboard"
        buttonTitle="Apply for a loan"
        style={{ flex: 1 }}
        actionFuntion={handleRedirectTOApplyForLoanScreen}
        isLoading={loading}
      >
        <ResponseGeneric
          header="No Loan Products Found"
          message="Please contact support for more details, You can go ahead and apply for a loan"
        />
      </ButtonContainer>
    );
  }

  return (
    <ButtonContainer
      title="Loan Application Dashboard"
      buttonTitle="Apply for a loan"
      style={{ flex: 1 }}
      actionFuntion={handleRedirectTOApplyForLoanScreen}
      isLoading={loading}
    >
      <View>
        {loanProducts &&
          loanProducts.map((item, i) => (
            <Card
              key={i}
              interest={item.interestRate}
              amount={item.maximumAmount}
              title={item.name}
              color={i % 2 === 0 ? theme.secondary : "white"}
              amountSign="$"
            />
          ))}
      </View>
    </ButtonContainer>
  );
};

export default Dashboard;
