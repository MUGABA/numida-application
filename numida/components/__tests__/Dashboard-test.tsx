import { ColorThemeProvider } from "@/hooks/context/colorContext";
import { gql } from "@apollo/client";
import { MockedProvider } from "@apollo/client/testing";
import {
  fireEvent,
  render,
  screen,
  waitFor,
} from "@testing-library/react-native";
import React from "react";
import Dashboard from "../Dashboard/dashboard";

const GET_LOAN_PRODUCTS = gql`
  query {
    loanProducts {
      id
      name
      interestRate
      maximumAmount
    }
  }
`;

const mocks = [
  {
    request: {
      query: GET_LOAN_PRODUCTS,
    },
    result: {
      data: {
        loanProducts: [
          {
            id: 1,
            name: "Tom's Loan",
            interestRate: 5.0,
            maximumAmount: 10000,
            __typename: "LoanProduct",
          },
        ],
      },
    },
  },
];

jest.mock("expo-router", () => ({
  useRouter: () => ({
    push: jest.fn(),
    replace: jest.fn(),
    goBack: jest.fn(),
    canGoBack: jest.fn(() => true),
  }),
  useSegments: jest.fn(() => ["home"]),
}));

const renderWithProvider = (ui: React.ReactElement) => {
  return render(<ColorThemeProvider>{ui}</ColorThemeProvider>);
};

describe("Dashboard Component", () => {
  it("Displays loan products from GraphQL", async () => {
    renderWithProvider(
      <MockedProvider mocks={mocks} addTypename={true}>
        <Dashboard />
      </MockedProvider>
    );
    await waitFor(() => expect(screen.getByText("Tom's Loan")).toBeTruthy());
  });

  it("Navigates to Loan Application screen on button click", async () => {
    renderWithProvider(
      <MockedProvider mocks={mocks} addTypename={true}>
        <Dashboard />
      </MockedProvider>
    );

    fireEvent.press(screen.getByText("Apply for a loan"));

    await waitFor(() => {
      expect(screen.getByText("Apply for a loan")).toBeTruthy();
    });
  });

  it(`renders correctly`, async () => {
    const { toJSON } = renderWithProvider(
      <MockedProvider mocks={mocks} addTypename={true}>
        <Dashboard />
      </MockedProvider>
    );

    await waitFor(() => {
      expect(toJSON()).toMatchSnapshot();
    });
  });
});
