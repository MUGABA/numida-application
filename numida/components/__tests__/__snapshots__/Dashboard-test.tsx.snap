// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dashboard Component renders correctly 1`] = `
<RCTScrollView>
  <View>
    <RNCSafeAreaView
      edges={
        {
          "bottom": "additive",
          "left": "additive",
          "right": "additive",
          "top": "additive",
        }
      }
      style={
        {
          "flex": 1,
        }
      }
    >
      <View
        onLayout={[Function]}
        style={
          {
            "alignItems": "center",
            "flex": 1,
            "justifyContent": "center",
          }
        }
      >
        <View>
          <View
            style={
              [
                {
                  "flex": 1,
                  "paddingHorizontal": 20,
                },
                {
                  "flex": 1,
                },
              ]
            }
          >
            <Text
              style={
                {
                  "color": "black",
                  "fontSize": 38,
                  "fontWeight": "bold",
                  "marginVertical": 10,
                  "paddingHorizontal": 12,
                  "paddingVertical": 25,
                }
              }
            >
              Loan Application Dashboard
            </Text>
            <View>
              <View />
            </View>
            <View>
              <View
                accessibilityState={
                  {
                    "busy": undefined,
                    "checked": undefined,
                    "disabled": undefined,
                    "expanded": undefined,
                    "selected": undefined,
                  }
                }
                accessibilityValue={
                  {
                    "max": undefined,
                    "min": undefined,
                    "now": undefined,
                    "text": undefined,
                  }
                }
                accessible={true}
                collapsable={false}
                focusable={true}
                onBlur={[Function]}
                onClick={[Function]}
                onFocus={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  {
                    "alignItems": "center",
                    "backgroundColor": "#30C2E3",
                    "borderRadius": 15,
                    "justifyContent": "center",
                    "marginTop": 80,
                  }
                }
              >
                <Text
                  style={
                    {
                      "borderRadius": 5,
                      "color": "white",
                      "fontSize": 18,
                      "paddingHorizontal": 8,
                      "paddingVertical": 12,
                      "textTransform": "uppercase",
                    }
                  }
                >
                  Apply for a loan
                </Text>
              </View>
            </View>
          </View>
        </View>
      </View>
    </RNCSafeAreaView>
    <View
      style={
        [
          {
            "alignItems": "center",
            "flex": 1,
            "height": 1334,
            "justifyContent": "center",
            "left": 0,
            "position": "absolute",
            "top": 0,
            "width": 750,
            "zIndex": 1000,
          },
          undefined,
        ]
      }
    >
      <ActivityIndicator
        color="#0000ff"
        size="large"
      />
    </View>
  </View>
</RCTScrollView>
`;
